src/types/dependency-container.ts

```diff
@@ -72,4 +74,28 @@ export default interface DependencyContainer {
 
   clearInstances(): void;
   createChildContainer(): DependencyContainer;
+
+  /**
+   * Registers a callback that is called when a specific injection token is resolved
+   * @param token The token to intercept
+   * @param callback The callback that is called before the token is resolved
+   * @param options Options for under what circumstances the callback will be called
+   */
+  beforeResolution<T>(
+    token: InjectionToken<T>,
+    callback: (t: InjectionToken<T>, r: ResolutionType) => void,
```

Comment:
Can you write interfaces for these callback types? That was we can easily have JSDocs for the params

```ts
interface BeforeResolutionCalback {
  /** ... */
  (token: ..., resolutionType: ...): void
}
```